Update Spec to Version 2.0 (#1)
AlamoEngine-Tools/eaw.modinfo: README.md

4. [`modinfo` Type Specification](#the-modinfo-type)
5. [`modreference` Type Specification](#the-modreference-type)
6. [`steamdata` Type Specification](#the-steamdata-type)
7. [Dependency Resolving](#dependency-resolving)
8. [Dendepency Test Cases](#dependency-resolving-test-cases)

## Filename

The meta information must be saved to a `JSON` file. There are two ways for naming the file:

1. The file is called `modinfo.json`. This is considered as the *main file*.
2. The file is called `[Any_FS_compliant_letter]-modinfo.json`. This is considered as a *variant file.*

Option `2` can be used if you want to create different variants of a mod that share the same files. 

*Imagine if your mod is a submod for not only one but two different mods. This way you only need to develop and upload the mod once but it can target both base mods simultaneously.* 

*Implementation Notes: In practice this will instanciate a new mod for each variant available.*

## File Position

The target directory is the top level of the mod's folder (next to where the mod's `data` folder is).

It may contain none, one or multiple files, as described in [Filename](#filename).

> If there exists variant files AND a main file the content from the main file gets merged into the variant file unless the variant overrides a property.  

Reviewer: ```suggestion
If there exists at least one variant file AND a main file the content from the main file gets merged into the variant file(-s) unless the variant overrides a property.  
```

If there are only variant files they each act as a main files on their own. 

> *Implementation Notes: As soon as a mod folder contains modinfo variant files, only these should yield an instance of a mod. The main modinfo file (if existing) or just the directry itself should get ignored.*

Reviewer: ```suggestion
*Implementation Notes: As soon as a mod folder contains modinfo variant files, only these should yield an instance of a mod. The main modinfo file (if existing) or just the directry itself should be ignored.*
```

### The `"icon"` Property

**Level:** *OPTIONAL*

**Data Type**: `String`  

**Data Semantics**: File System Path 

**Description:**

The path to the mod's icon file **relative** to the mod's root directory or an **absolute** path.

### The `"version"` Property

**Level:** *OPTIONAL*

**Data Type**: `String`  

**Data Semantics**: 4 Digit Semantic Version 

**Description:**

The mod's version according to the extended semantic versioning: [Semantic Versioning](https://semver.org/) that also supports a fourth digits for build increments, etc. and suffixes (e.g. `"-rc1"`).

### The `"dependencies"` Property

**Level:** *OPTIONAL*

**Data Type**: [`modreference`](#the-modreference-type)`[]`

**Data Semantics**: Ordered List of Objects 

**Description:**

The `dependencies` container holds an ordered list of [`"modreference type"`](#the-modreference-type)s this mod relies on.

The list is either absent from the `modinfo.json` or contains at least one item.

The first mod in the list resembles the closest dependeny. Every `n+1` item is another dependeny of the target mod.
If a dependency has dependencies itself these **must not** be added to the list.

> **The mod of the this modinfo.json file must not be listed here because that would result in a cycle!**

Reviewer: So the mod described by the file itsef?
```suggestion
**The mod described by the current `modinfo.json` file must not be listed here, including it would result in a cycle!**
```

```
  C
 /
A - D
 \
  B
```

### The `"steamdata"` Property

**Level:** *OPTIONAL*

**Data Type**: [`steamdata`](#tthe-steamdata-type)  

**Data Semantics**: Steam Workshops JSON

**Description:**

The [`"steamdata type"`](#the-steamdata-type) container holds additional info that is required for the Steam Version of the game.
The container is either absent from the `modinfo.json` or it is fully required.

### The `"custom"` Property

**Level:** *OPTIONAL*

**Data Type**: `Object[]`

**Data Semantics**: A list of anything

**Description:**

The `custom` property allows arbitrary extensions to the format. Programs implementing the core format should always be able to serialize the whole object, but the custom data wrapped within the extension object has to be accounted for only where applicable and/or needed.

---

## The `"modreference"` Type

#### The `"modreference.modtype"` Property

**Level:** **REQUIRED**

**Data Type**: `integer`

**Data Semantics**: Enum

**Description:**

The modtype enumeration:

| Value | Meaning |
|:--:|:--|
|`0`|any normal mod inside the Mods/ directory|
|`1`|a Steam Workshops mod|
|`2`|a "virtual" mod. (Currently unsupported)|

> *Reasonable: The current mod does NOT contain a `modtype` property because the mod should not know it's owen type. Otherwise sharing this file across steam and disk mods would not be possible. A `modreference` requests this data, meaning tool support to determine the actual `modtype` is necessary. This design coice was made because mod linking should always be considered for Steam Workshop mods. For convenient test for mod creators we keep the possibility to reference to local mods.*

Reviewer: ```suggestion
*Rationale: The current mod does NOT contain a `modtype` property because the mod should not have to know it's own type. Otherwise sharing this file across steam and disk mods would not be possible. A `modreference` requests this data, meaning tool support to determine the actual `modtype` is necessary. This design coice was made because mod linking should always be considered for Steam Workshop mods. The possibility to reference local mods is a convenience functionality intended to be used by mod developers for test setups and development.*
```

> *Becuase a workshop mod in theroy also is a `default` mod the number values are chooses the way they are.*

Reviewer: ```suggestion
*Because a workshop mod in theroy also is the `default` mod type the numeric enumeration values are choosen the way they are.*
```

**Description:**

This property either contains an absolute or relative path of the parent mod or holds the STEAMID for workshop mods.

---

## The `"steamdata"` Type

### The `"steamdata.publishedfileid"` Property

**Level:** **REQUIRED**

**Data Type**: `string`

**Data Semantics**: Unique identifier

**Description:**

The Steam Workshop ID.

### The `"steamdata.contentfolder"` Property

**Level:** **REQUIRED**

**Data Type**: `String`

**Data Semantics**: File System Path

**Description:**

The content folder's name as specified by the Steam Uploader.

### The `"steamdata.visibility"` Property

**Level:** **REQUIRED**

**Data Type**: `integer`

**Data Semantics**: Enum

**Description:**

The visibility enumeration:

| Value | Meaning |
|:--:|:--|
|`0`|hidden|
|`1`|friends only|
|`2`|public|

### The `"steamdata.metadata"` Property

**Level:** **REQUIRED**

**Data Type**: `String`

**Data Semantics**: Text

**Description:**

Arbitrary metadata as string.

### The `"steamdata.tags"` Property

**Level:** **REQUIRED**

**Data Type**: `String[]`

**Data Semantics**: Collection of tags

**Description:**

Steam Tags as specified by the Steam Uploader.

> At least either `EAW` or `FOC` is required to determine the game the mod shows up for.

Reviewer: ```suggestion
At least the tag `EAW` or the tag `FOC` is required to determine the game the mod is made for which also controls on which Steam Workshop the mod is listed..
A mod may **never** use both tags at the same time, they are mutually exclusive.
```

## Dependency Resolving
The game supports chaining (could also be called *overriding* or *linked*) mods by queuing up the command line arguments `STEAMMOD` and/or `MODPATH`. 

> While the Command Line options only can resemble a line (or more precisely a Queue), real mod dependencies can look like complex tree due to multiple inheritance. Thus there needs to be a deterministic logic to convert the real hierachy into a CLI compatible representation. 

Reviewer: ```suggestion
While the Command Line options only can resemble a line (or more precisely a Queue), real mod dependencies can look like complex tree structures due to multiple inheritance. Thus there needs to be a deterministic logic to convert the actual hierachy into a CLI compatible representation. 
```

To describe relations between mods this specification introduces an optional [`dependency`](#the-dependencies-property) property which is an ordered list. Each item in the list branches from current mod, which fullfils the multiple inheritance logic.

The conversion from a tree structure to a line is called *flattening*.
Implementations of this standard must fullfill the following requirements to create a sucessfull conversion:

1. The `dependencies` list MUST be processed from start to end.

> 2. Before processing dependiencies of `n+1` (where n is the current processed mod) all dependencies of `n` must be processed first.

Reviewer: ```suggestion
2. Before processing dependiencies of `n+1` (where `n` is the mod that is being processed at the time) all dependencies of `n` must be processed first.
```

> 3. Duplicates MUST get removed. *Having duplicates that are not direct neighbours is a strong indicator you have a dependency cycle!*

Reviewer: ```suggestion
3. Duplicates MUST be removed. *Having duplicates that are not direct neighbours is a strong indicator for a dependency cycle!*
```

> 4. The implementation MUST be able to recognice and react on dependency cycles. 

Reviewer: ```suggestion
4. The implementation MUST be able to recognise and respond to identified dependency cycles accordingly. 
```

> 5. The provided [test cases](#dependency-resolving-test-cases) MUST pass:

Reviewer: ```suggestion
5. The provided [test cases](#dependency-resolving-test-cases) MUST pass.
```

6. The result of the *flattening* MAY get reversed depending on the requirements of the consumer.

*Note: Converting a (non-binary) tree structure to a flatteded and duplicate-free line is a one-way operation. It also looses accuracy. The requirements above are created to ensure consistency across multiple implementations.*

> *Advise: For making sure you mod always work as intended, we recommend to keep the inheritance level to a minimum. Also try to avoid multiple inheritances.*

Reviewer: ```suggestion
*Advise: For making sure your mod always works as intended, we recommend to keep the inheritance level to a minimum. Also try to avoid multiple inheritances.*
```

### Dependency Resolving Test Cases

> Node `A` is always the mod that should get executed. Everything behind  `:` means these are the direct dependencies.

Reviewer: ```suggestion
Node `A` is always the mod that should be loaded. Every mod following after  `:` are direct dependencies of the mod, if multiple they are separated by a comma.
```
